libbeat/common/event.go

```diff
@@ -1,92 +1,223 @@
 package common
 
 import (
+	"encoding"
 	"encoding/json"
+	"fmt"
 	"reflect"
-	"time"
+	"strconv"
+	"strings"
 
 	"github.com/elastic/beats/libbeat/logp"
+
+	"github.com/pkg/errors"
 )
 
-func MarshallUnmarshall(v interface{}) (MapStr, error) {
-	// decode and encode JSON
-	marshaled, err := json.Marshal(v)
-	if err != nil {
-		logp.Warn("marshal err: %v", err)
-		return nil, err
+const eventDebugSelector = "event"
+
+var eventDebugf = logp.MakeDebug(eventDebugSelector)
+
+var textMarshalerType = reflect.TypeOf((*encoding.TextMarshaler)(nil)).Elem()
+
+// ConvertToGenericEvent normalizes the types contained in the given MapStr.
+//
+// Nil values in maps are dropped during the conversion. Any unsupported types
+// that are found in the MapStr are dropped and warnings are logged.
+func ConvertToGenericEvent(m MapStr) MapStr {
+	event, errs := normalizeMap("", m)
+	if len(errs) > 0 {
+		logp.Warn("Unsuccessful conversion to generic event: %v errors: %v, "+
+			"event=%#v", len(errs), errs, m)
 	}
-	var v1 MapStr
-	err = json.Unmarshal(marshaled, &v1)
-	if err != nil {
-		logp.Warn("unmarshal err: %v", err)
-		return nil, err
+	return event
+}
+
+// normalizeMap normalizes each element contained in the given map. If an error
+// occurs during normalization, processing of m will continue, and all errors
+// are returned at the end.
+func normalizeMap(baseKey string, m MapStr) (MapStr, []error) {
+	var errs []error
+
+	out := make(MapStr, len(m))
+	for key, value := range m {
+		fullKey := joinKeys(baseKey, key)
```

Comment:
Yeah, it's only for error reporting. I made some change to minimize the number of allocations associated with this feature. And now it only concatenates the keys if an error occurs.